doc: mention that intrinsics should not be called in user code

scrabsha opened the pull request:
Intrinsic functions declared in `std::intrinsics` are an implementation detail and should not be called directly by the user. The compiler explicitly warns against their use in user code:

```
warning: the feature `core_intrinsics` is internal to the compiler or standard library
 --> src/lib.rs:1:12
  |
1 | #![feature(core_intrinsics)]
  |            ^^^^^^^^^^^^^^^
  |
  = note: using it is strongly discouraged
  = note: `#[warn(internal_features)]` on by default
```

[**Playground link**]

This PR documents what the compiler warning says: these intrinsics should not be used in user code.

[**Playground link**]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=1c893b0698291f550bbdde0151fd221b

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

workingjubilee commented on library/core/src/intrinsics/mod.rs:
I'm slightly curious about the motivation. Would it make more sense to simply put this at the top of the file, as the module documentation?

RalfJung replied:
Yeah, agreed... repeating this everywhere is not great and makes this already big file even bigger.

scrabsha committed:
doc: mention that intrinsics should not be called in user code

Intrinsic functions declared in `std::intrinsics` are an implementation
detail and should not be called directly by the user. The compiler
explicitly warns against their use in user code:

```
warning: the feature `core_intrinsics` is internal to the compiler or standard library
 --> src/lib.rs:1:12
  |
1 | #![feature(core_intrinsics)]
  |            ^^^^^^^^^^^^^^^
  |
  = note: using it is strongly discouraged
  = note: `#[warn(internal_features)]` on by default
```

[**Playground link**]

This PR documents what the compiler warning says: these intrinsics should
not be called outside the standard library.

[**Playground link**]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=1c893b0698291f550bbdde0151fd221b

RalfJung commented on library/core/src/intrinsics/mod.rs:
```diff
@@ -6,6 +6,9 @@
 //! and for const evaluation in <https://github.com/rust-lang/rust/blob/master/compiler/rustc_const_eval/src/interpret/intrinsics.rs>.
 //!
+//! The functions in this module are implementation details of `core` and should
+//! not be used outside of the standard library.
```
Maybe moved this up even further so it is the first paragraph after the "title"?

lcnr replied:
:+1: 
```rust
//! Compiler intrinsics.
//!
//! These are the imports making intrinsics available to Rust code. The actual
//! implementations live in the compiler.  The functions in this module are
//! implementation details of `core` and should not be used outside of the
//! standard library. We generally provide access to intrinsics via stable
//! wrapper functions. Use these instead.
```
something like this?

scrabsha replied:
Speaking of which, should I use _should_ or _must_?

I can't think of a reasonable use of an intrinsic outside the standard library, but I'd like to avoid accidentally blocking people.

lcnr replied:
:shrug: must feels a bit too strong, it's totally fine to use these intrinsics we just provide no guarantees about them.

RalfJung commented:
Looks good, thanks!

@bors r+ rollup

bors commented:
:pushpin: Commit 527f35a28f2661e40b1b6824103861312f3a188d has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung 527f35a28f2661e40b1b6824103861312f3a188d -->
<!-- homu: {"type":"Approved","sha":"527f35a28f2661e40b1b6824103861312f3a188d","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
